bpf/generictracer/protocol_mysql.h

```diff
@@ -0,0 +1,664 @@
+#pragma once
+
+#include <bpfcore/vmlinux.h>
+#include <bpfcore/bpf_helpers.h>
+
+#include <common/common.h>
+#include <common/connection_info.h>
+#include <common/http_types.h>
+#include <common/pin_internal.h>
+#include <common/ringbuf.h>
+#include <common/runtime.h>
+#include <common/sql.h>
+#include <common/tp_info.h>
+#include <common/trace_common.h>
+
+#include <generictracer/protocol_common.h>
+
+#include <maps/active_ssl_connections.h>
+
+// Every mysql command packet is prefixed by an header
+// https://mariadb.com/kb/en/0-packet/
+struct mysql_hdr {
+    uint32_t payload_length : 24;
+    uint8_t sequence_id;
+    uint8_t command_id;
+
+    // Metadata
+    bool hdr_arrived; // Signals whether to skip or not the first 4 bytes in the current buffer as
+                      // they arrived in a previous packet.
+} __attribute__((packed));
+
+struct mysql_state_data {
+    uint32_t payload_length : 24;
+    uint8_t sequence_id;
+} __attribute__((packed));
+
+struct {
+    __uint(type, BPF_MAP_TYPE_LRU_HASH);
+    __type(key, connection_info_t);
+    __type(value, struct mysql_state_data);
+    __uint(max_entries, MAX_CONCURRENT_REQUESTS);
+} mysql_state SEC(".maps");
+
+#define MYSQL_HDR_SIZE 5
+#define MYSQL_HDR_COMMAND_ID_SIZE 1
+#define MYSQL_HDR_SEQUENCE_ID_SIZE 1
+#define MYSQL_HDR_PAYLOAD_LENGTH_SIZE 3
+#define MYSQL_HDR_WITHOUT_COMMAND_SIZE 4
+#define MYSQL_ERR_PACKET_MIN_SIZE 9
+
+// MySQL Client
+#define MYSQL_COM_QUERY 0x3
+#define MYSQL_COM_STMT_PREPARE 0x16
+#define MYSQL_COM_STMT_EXECUTE 0x17
+// MySQL Server
+#define MYSQL_OK 0x0
+#define MYSQL_OK_EOF_DEPRECATED 0xfe
+#define MYSQL_ERR 0xff
+
+#define MYSQL_STATE_MARKER '#'
+
+static inline void mysql_store_state_data(connection_info_t *conn_info, const unsigned char *data) {
+    struct mysql_state_data *state_data = bpf_map_lookup_elem(&mysql_state, conn_info);
+    if (state_data == NULL) {
+        struct mysql_state_data new_state_data = {};
+        bpf_probe_read(&new_state_data, MYSQL_HDR_PAYLOAD_LENGTH_SIZE, (void *)data);
+        bpf_probe_read(&new_state_data.sequence_id, MYSQL_HDR_SEQUENCE_ID_SIZE, (void *)(data + 3));
+        bpf_map_update_elem(&mysql_state, conn_info, &new_state_data, BPF_ANY);
+    } else {
+        bpf_probe_read(state_data, MYSQL_HDR_WITHOUT_COMMAND_SIZE, (void *)data);
+        bpf_map_update_elem(&mysql_state, conn_info, state_data, BPF_ANY);
+    }
+}
+
+static inline int mysql_parse_fixup_header(connection_info_t *conn_info,
```

Comment:
```suggestion
static inline int mysql_parse_fixup_header(const connection_info_t *conn_info,
```

having said that, any chance you can split this function into smaller bits? I feel it's doing too many things (i.e. deleting data _sometimes_, and what not.